# <center>💟INFORME MACHING LEARNING💟</center>

El objetivo de este informe es crear dos notebooks. Uno donde se utilice la técnica de redes neuronales
y otro para la técnica de árboles de decisión, todo esto para determinar los hiperparámetros óptimos para obtener la mayor precisión posible.

💟INTEGRANTES💟
* Isabela Rosero 2128720
* Luisa Cardenas 1823494 
* Stefhania Noguera 2125854

💟COMO CORRER EL ARCHIVO💟
* Instalar pandas con *pip install pandas*
* Instalar sklearn con *pip install  scikit-learn*

💟IMPORTACIONES💟

In [5]:
# Importar bibliotecas necesarias
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

💟LEER EL ARCHIVO CSV💟

In [6]:
# 1. Leer el archivo breastCancer.csv.
data = pd.read_csv('breastCancer.csv')

💟SELECCIONAR EL 80% DEL CONJUNTO DE DATOS Y EL 20 PARA PRUEBAS💟

In [7]:
# 2. Seleccionar aleatoriamente el 80% del conjunto de datos para entrenar y el 20% restante para las pruebas.
train_data, test_data = train_test_split(data, test_size=0.2, random_state=123)

💟NORMALIZAR LOS DATOS💟

In [8]:
# 3. Utilizar una estrategia para normalizar los datos.
scaler = StandardScaler()
X_train = scaler.fit_transform(train_data.iloc[:, :-1])  # Excluir la columna de diagnóstico
y_train = train_data['diagnosis']

X_test = scaler.transform(test_data.iloc[:, :-1])
y_test = test_data['diagnosis']

💟CONSTRUIR 5 REDES CON VARIACIONES💟

In [10]:
# 4. Construir 5 redes neuronales variando la función de activación, el solver, y la cantidad de capas ocultas y de neuronas por cada capa oculta.
results = []

# Seleccionar combinaciones específicas
combinations = [
    ('relu', 'adam', (200 , 300)),
    ('tanh', 'lbfgs', (10, 20, 10)),
    ('logistic', 'sgd', (20, 15, 10)),
    ('identity', 'adam', (10, 10, 5, 5)),
    ('tanh', 'adam', (150, 75)),
]

for activation, solver, hidden_layer_size in combinations:
    # Crear y entrenar el modelo
    model = MLPClassifier(activation=activation, solver=solver, hidden_layer_sizes=hidden_layer_size, random_state=123, max_iter=500)
    model.fit(X_train, y_train)

    # Hacer predicciones
    y_pred = model.predict(X_test)

    # Calcular accuracy y matriz de confusión
    acc = accuracy_score(y_test, y_pred)
    conf_matrix = confusion_matrix(y_test, y_pred)

    # Guardar resultados
    results.append({
        'Activation': activation,
        'Solver': solver,
        'Hidden Layer Sizes': hidden_layer_size,
        'Accuracy': acc,
        'Confusion Matrix': conf_matrix
    })

# Mostrar resultados en una tabla
results_df = pd.DataFrame(results)
print(results_df)


  Activation Solver Hidden Layer Sizes  Accuracy    Confusion Matrix
0       relu   adam         (200, 300)  0.991228  [[73, 0], [1, 40]]
1       tanh  lbfgs       (10, 20, 10)  0.938596  [[70, 3], [4, 37]]
2   logistic    sgd       (20, 15, 10)  0.640351  [[73, 0], [41, 0]]
3   identity   adam     (10, 10, 5, 5)  0.973684  [[73, 0], [3, 38]]
4       tanh   adam          (150, 75)  0.964912  [[72, 1], [3, 38]]


c:\Users\User\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


💟Variaciones con hyperparametro alpha💟


In [11]:
print("")
print("Se varia el hyperparametro de alpha")
# Entrenar primera red con alpha=0.3
model_1 = MLPClassifier(activation='relu', solver='adam', hidden_layer_sizes=(200, 300), alpha=0.3, random_state=123, max_iter=500)
model_1.fit(X_train, y_train)

# Hacer predicciones para la primera red
y_pred_1 = model_1.predict(X_test)

# Calcular accuracy y matriz de confusión para la primera red
acc_1 = accuracy_score(y_test, y_pred_1)
conf_matrix_1 = confusion_matrix(y_test, y_pred_1)

# Guardar resultados de la primera red
results_1 = {
    'Activation': 'relu',
    'Solver': 'adam',
    'Hidden Layer Sizes': (200, 300),
    'Alpha': 0.3,
    'Accuracy': acc_1,
    'Confusion Matrix': conf_matrix_1
}

# Entrenar segunda red con alpha=0.1
model_2 = MLPClassifier(activation='relu', solver='adam', hidden_layer_sizes=(200, 300), alpha=0.1, random_state=123, max_iter=500)
model_2.fit(X_train, y_train)

# Hacer predicciones para la segunda red
y_pred_2 = model_2.predict(X_test)

# Calcular accuracy y matriz de confusión para la segunda red
acc_2 = accuracy_score(y_test, y_pred_2)
conf_matrix_2 = confusion_matrix(y_test, y_pred_2)

# Guardar resultados de la segunda red
results_2 = {
    'Activation': 'relu',
    'Solver': 'adam',
    'Hidden Layer Sizes': (200, 300),
    'Alpha': 0.1,
    'Accuracy': acc_2,
    'Confusion Matrix': conf_matrix_2
}

# Mostrar resultados en una tabla
results_df = pd.DataFrame([results_1, results_2])
print(results_df)


Se varia el hyperparametro de alpha
  Activation Solver Hidden Layer Sizes  Alpha  Accuracy    Confusion Matrix
0       relu   adam         (200, 300)    0.3  0.964912  [[72, 1], [3, 38]]
1       relu   adam         (200, 300)    0.1  0.982456  [[72, 1], [1, 40]]


# <center>💟CODIGO COMPLETO💟</center>


In [4]:
# Importar bibliotecas necesarias
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# 1. Leer el archivo breastCancer.csv.
data = pd.read_csv('breastCancer.csv')

# 2. Seleccionar aleatoriamente el 80% del conjunto de datos para entrenar y el 20% restante para las pruebas.
train_data, test_data = train_test_split(data, test_size=0.2, random_state=123)

# 3. Utilizar una estrategia para normalizar los datos.
scaler = StandardScaler()
X_train = scaler.fit_transform(train_data.iloc[:, :-1])  # Excluir la columna de diagnóstico
y_train = train_data['diagnosis']

X_test = scaler.transform(test_data.iloc[:, :-1])
y_test = test_data['diagnosis']

# 4. Construir 5 redes neuronales variando la función de activación, el solver, y la cantidad de capas ocultas y de neuronas por cada capa oculta.
results = []

# Seleccionar combinaciones específicas
combinations = [
    ('relu', 'adam', (200 , 300)),
    ('tanh', 'lbfgs', (10, 20, 10)),
    ('logistic', 'sgd', (20, 15, 10)),
    ('identity', 'adam', (10, 10, 5, 5)),
    ('tanh', 'adam', (150, 75)),
]

for activation, solver, hidden_layer_size in combinations:
    # Crear y entrenar el modelo
    model = MLPClassifier(activation=activation, solver=solver, hidden_layer_sizes=hidden_layer_size, random_state=123, max_iter=500)
    model.fit(X_train, y_train)

    # Hacer predicciones
    y_pred = model.predict(X_test)

    # Calcular accuracy y matriz de confusión
    acc = accuracy_score(y_test, y_pred)
    conf_matrix = confusion_matrix(y_test, y_pred)

    # Guardar resultados
    results.append({
        'Activation': activation,
        'Solver': solver,
        'Hidden Layer Sizes': hidden_layer_size,
        'Accuracy': acc,
        'Confusion Matrix': conf_matrix
    })

# Mostrar resultados en una tabla
results_df = pd.DataFrame(results)
print(results_df)
print("")
print("Se varia el hyperparametro de alpha")

#Variaciones con hyperparametro alpha

# Entrenar primera red con alpha=0.3
model_1 = MLPClassifier(activation='relu', solver='adam', hidden_layer_sizes=(200, 300), alpha=0.3, random_state=123, max_iter=500)
model_1.fit(X_train, y_train)

# Hacer predicciones para la primera red
y_pred_1 = model_1.predict(X_test)

# Calcular accuracy y matriz de confusión para la primera red
acc_1 = accuracy_score(y_test, y_pred_1)
conf_matrix_1 = confusion_matrix(y_test, y_pred_1)

# Guardar resultados de la primera red
results_1 = {
    'Activation': 'relu',
    'Solver': 'adam',
    'Hidden Layer Sizes': (200, 300),
    'Alpha': 0.3,
    'Accuracy': acc_1,
    'Confusion Matrix': conf_matrix_1
}

# Entrenar segunda red con alpha=0.1
model_2 = MLPClassifier(activation='relu', solver='adam', hidden_layer_sizes=(200, 300), alpha=0.1, random_state=123, max_iter=500)
model_2.fit(X_train, y_train)

# Hacer predicciones para la segunda red
y_pred_2 = model_2.predict(X_test)

# Calcular accuracy y matriz de confusión para la segunda red
acc_2 = accuracy_score(y_test, y_pred_2)
conf_matrix_2 = confusion_matrix(y_test, y_pred_2)

# Guardar resultados de la segunda red
results_2 = {
    'Activation': 'relu',
    'Solver': 'adam',
    'Hidden Layer Sizes': (200, 300),
    'Alpha': 0.1,
    'Accuracy': acc_2,
    'Confusion Matrix': conf_matrix_2
}

# Mostrar resultados en una tabla
results_df = pd.DataFrame([results_1, results_2])
print(results_df)


c:\Users\User\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


  Activation Solver Hidden Layer Sizes  Accuracy    Confusion Matrix
0       relu   adam         (200, 300)  0.991228  [[73, 0], [1, 40]]
1       tanh  lbfgs       (10, 20, 10)  0.938596  [[70, 3], [4, 37]]
2   logistic    sgd       (20, 15, 10)  0.640351  [[73, 0], [41, 0]]
3   identity   adam     (10, 10, 5, 5)  0.973684  [[73, 0], [3, 38]]
4       tanh   adam          (150, 75)  0.964912  [[72, 1], [3, 38]]

  Activation Solver Hidden Layer Sizes  Alpha  Accuracy    Confusion Matrix
0       relu   adam         (200, 300)    0.3  0.964912  [[72, 1], [3, 38]]
1       relu   adam         (200, 300)    0.1  0.982456  [[72, 1], [1, 40]]


# <center>💟TABLA DE RESULTADOS💟</center>


|💟Layer Sizes💟|💟Accuracy💟|💟Confusion Matrix💟|
|--|--|--|
|0     (200, 300) |0.991228| [73, 0], [1, 40]|
|1     (10, 20, 10) | 0.938596 | [70, 3], [4, 37]|
|2  (20, 15, 10) | 0.640351 | [73, 0], [41, 0]|
|3  (10, 10, 5, 5) | 0.973684 | [73, 0], [3, 38]|
|4  (150, 75) | 0.964912 | [72, 1], [3, 38]|

💟Hiperparámetros que por el momento le permiten obtener la red con mayor accuracy💟
*   Con el solver "adam"
*   Con una activacion "relu"
*   Con un random state de 123

Despues se modificó el hiperparámetro 💟alpha💟 

Se modifico con 0.3 y 0.1 llegando a la conclusion de:

💟El mejor accuracy lo tuvo en: 0.1💟